napari/_vispy/vispy_image_layer.py

```diff
@@ -57,7 +57,18 @@ def _on_display_change(self, data=None):
         self.node.order = self.order
         self.reset()
 
+    def _data_astype(self, data, dtype):
+        """Broken out as a separate function so we can time with perfmon."""
+        return data.astype(dtype)
+
     def _on_data_change(self, event=None):
+        """Our self.layer._data_view has been updated, update our node.
+        """
+        if not self.node.loaded:
```

Comment:
Is there a reason you can't just use `self.layer.loaded` here too? I don't really like having `loaded` on the `node` as we're going to want to remove that vendored copy of the vispy Image Visual once vispy makes their next release, and I don't think we are likely to push the `loaded` variable upstream to them any time soon. We could just have it on this object instead if we need.

Reply:
I changed it to self.layer. This used to be "visible" not "loaded" and when I switched it to loaded I didn't make this change. However there is still a "loaded" in the vendored image/volume because it's checked in `VispyImageLayer._on_data_change`. There might be some other way to do that, but that's how it's always worked in #1510 (although it changed from visible to loaded recently).

So I think the plan is finished splitting #1510, then add the octree visual version 0 which will get rid of this "loaded" flag. Instead we'll have an always visible container of some kind that manages its subchildren. That's where we need to be for many reasons.

Reply:
> However there is still a "loaded" in the vendored image/volume because it's checked in VispyImageLayer._on_data_change. There might be some other way to do that, but that's how it's always worked in #1510 (although it changed from visible to loaded recently).

Looking at the code in #1510 and the code now in this PR I really do think you could drop the entire `loaded` property from the visual right now if you wanted. I see how it gets used inside `set_data` on the visual to trigger a return, but we control when that the `set_data` call happens in our `_on_data_change` and `_on_display_change` methods and so can check `self.layer.loaded` before making the call from both those functions (right now you already check from `_on_data_change`). It doesn't look like there are additional calls to `set_data` that can get triggered internally by vispy, and I'm not quite sure how they would use the method if they wanted, but I guess that might be one thing to verify. If not, then having the checks in our code would mean you wouldn't need the check in `set_data` anymore and could drop the `loaded` property entirely from the visual.

I did just give it a try to delete `loaded` from the node and comment out the `self.node.loaded` references and async still seemed to worked as before, so it is probably pretty harmless to make this change and would mean that we'd have a much easier upgrade path after the next vispy release.

> So I think the plan is finished splitting #1510, then add the octree visual version 0 which will get rid of this "loaded" flag. Instead we'll have an always visible container of some kind that manages its subchildren. That's where we need to be for many reasons.

That sounds like a good plan, and I won't hold up the PR for this change, but I do think dropping `loaded` from our vispy image layer object will be easier than dropping from the vendored vispy visual, which really shouldn't be in the napari code base and is only there due to the lack of a recent vispy release.

Reply:
Okay yes it works your way, I changed it. It's better. I did it the other way basically just because checking inside "set_data" I knew we were 100% covered and it's pretty hard to search for "all places we trigger than event". But it seems to work how you say and yes it's much better to not change the vendored files.

But even the checks should go away with the octree visuals version 0, hopefully.

Reply:
Great, thanks @pwinston much appreciated